### DISCUSSION TITLE
Window implementation

### DISCUSSION THREAD

**Discussion by blaisebarrette** [HomeSpan/HomeSpan#336] [Q&A] [answered]
Hi All!

First, I would like to say that I am very impressed with the ease of use and adaptability of this library.
I have made a Ceiling fan control and it's working flawlessly so far!

I moved to the next project witch will be a control for windows opening and closing. Mechanically, it's all setup. Time to look at the electronics.

I've setup a bench prototype and am now working on the code.

I can't see what's missing. It's uploading fine to the esp32 but it's not showing up in HomeKit.

The "Update" and the "Loop" part of the code is not fully coded yet but I was hoping to have enough code to pair with HomeKit and start experimenting.

Thanks for your help!

Her is what I have so far:

```
#include "HomeSpan.h" 

struct WindowOpener : Service::Window { // Window Opener
    int MotorPin1, MotorPin2;
    SpanCharacteristic *Current;        // reference to the Current window State Characteristic
    SpanCharacteristic *Target;         // reference to the Target window State Characteristic 
    SpanCharacteristic *State;          // reference to the position state Characteristic
    SpanCharacteristic *Hold;           // reference to the hold position Characteristic
    SpanCharacteristic *Obstruction;    // reference to the Obstruction Detected Characteristic

    WindowOpener(int MotorPin1, int MotorPin2) : Service::Window(){ // constructor() method
        
    Current=new Characteristic::CurrentPosition(0);          // initial value of 0 means closed
    Target=new Characteristic::TargetPosition(0);            // initial value of 0 means closed
    State=new Characteristic::PositionState(2);              // initial value of 1 means closed
    Hold=new Characteristic::HoldPosition(false);                // initial value of 1 means closed
    Obstruction=new Characteristic::ObstructionDetected(false);  // initial value of 1 means closed
        
    Serial.print("Configuring Window Opener");   // initialization message
    Serial.print("\n");

    this->MotorPin1=MotorPin1;
    this->MotorPin2=MotorPin2;

    pinMode(MotorPin1,OUTPUT);
    pinMode(MotorPin2,OUTPUT);

    digitalWrite(MotorPin1,HIGH);
    digitalWrite(MotorPin2,HIGH);

} // end constructor

  boolean update(){ // update() method

    // see HAP Documentation for details on what each value represents
    if(Target->getNewVal()==0){         // if the target-state value is set to 0, HomeKit is requesting the door to be in open position
      LOG1("Opening Window\n");
      Current->setVal(2);               // set the current-state value to 2, which means "opening"
      digitalWrite(MotorPin1,HIGH);
      digitalWrite(MotorPin2,LOW);
      Obstruction->setVal(false);       // clear any prior obstruction detection
    } else {
      LOG1("Closing Window\n");         // else the target-state value is set to 1, and HomeKit is requesting the door to be in the closed position
      Current->setVal(3);               // set the current-state value to 3, which means "closing"  
      digitalWrite(MotorPin1,LOW);
      digitalWrite(MotorPin2,HIGH);       
      Obstruction->setVal(false);       // clear any prior obstruction detection
    }

    if(State->getNewVal()==0){

    }
    
    return(true);                       // return true
  
  } // update

  void loop(){  // loop() method

    if(Current->getVal()==Target->getVal()) // if current-state matches target-state there is nothing do -- exit loop()
    digitalWrite(MotorPin1,LOW);
    digitalWrite(MotorPin2,LOW);
      return;

    if(Current->getVal()==4)                // if the current-state is stopped, there is nothing more to do - exit loop()     
    digitalWrite(MotorPin1,LOW);
    digitalWrite(MotorPin2,LOW);
      return;

    Current->setVal(Target->getVal());      // set the current-state to the target-state
       
  } // loop
  
};


void setup() {
  Serial.begin(115200);
  homeSpan.setStatusPin(2);
  homeSpan.setPairingCode("46637727");
  homeSpan.setSketchVersion("1.0.0");
  homeSpan.setQRID("WNDO");
  homeSpan.enableOTA(); // password is "homespan-ota"
  homeSpan.begin(Category::Windows,"Windows control");
  new SpanAccessory();
    new Service::AccessoryInformation();
      new Characteristic::Identify(); 
    new WindowOpener(16,17); // Create a Window Opener (see DEV_DoorsWindows.h for definition)
} // end of setup()

void loop(){
  homeSpan.poll();
} // end of loop()
```

**Comment by blaisebarrette** [comment]
Never mind... The code works. The esp32 had been used for a different HomeSpan project. It just needed to be erased to be pared again.

**Comment by HomeSpan** [comment, marked as answer]
Glad it's now working!

When re-using the same device for multiple HomeSpan projects it's often a good idea to refresh the device ID by typing 'H' into the CLI.  This retains all WiFi information, and does not alter the pairing setup code, but does cause HomeSpan to  generate a new HomeKit Device ID upon rebooting.  Device IDs are remembered by the Home App - if you change the Accessories configuration on a HomeSpan device, the Home App can often become "confused" and either render your Accessory incorrectly, or fail to show it at all.  In my experience the Home App may correct itself after while, but clearing the Device ID with the 'H' command (and then re-pairing) avoids the issue altogether.

**Reply by blaisebarrette** [reply]
OK thanks! Good to know! I was using the 'E' command and re-entering the Wifi credentials. It worked but the H command will save me time! 

Thanks!

**Comment by blaisebarrette** [comment]
I've just hit a new snag...

My code seems to work really nicely! Motors are turning in the right direction for three seconds as expected from my code.
All that's left to do is implementing the limit switches which should be a breeze.

The current bug I'm getting is that I can instantiate three windows and all works great but as soon as I instantiate a fourth one, the esp32 keeps on rebooting.

Here is the code:

```
#include "HomeSpan.h"

struct WindowOpener : Service::Window { // Window Opener
    int MotorPin1, MotorPin2;
    SpanCharacteristic *Current;        // reference to the Current window State Characteristic
    SpanCharacteristic *Target;         // reference to the Target window State Characteristic 
    SpanCharacteristic *State;          // reference to the position state Characteristic
    SpanCharacteristic *Hold;           // reference to the hold position Characteristic

    WindowOpener(int MotorPin1, int MotorPin2) : Service::Window(){ // constructor() method

    Current=(new Characteristic::CurrentPosition(0))->setRange(0,100,100);        // initial value of 0 means closed, 100 meand fully open
    Target=(new Characteristic::TargetPosition(0))->setRange(0,100,100);            // initial value of 0 means closed, 100 meand fully open
    State=new Characteristic::PositionState(2);              // initial value of 2 means stoped, 0 means closing, 1 means opening
    Hold=new Characteristic::HoldPosition(false);                // initial value of 1 means hold position, 0 is ignored
        
    Serial.print("Configuring Window Opener");   // initialization message
    Serial.print("\n");

    this->MotorPin1=MotorPin1;
    this->MotorPin2=MotorPin2;

    pinMode(MotorPin1,OUTPUT);
    pinMode(MotorPin2,OUTPUT);

    digitalWrite(MotorPin1,LOW);
    digitalWrite(MotorPin2,LOW);

} // end constructor

  boolean update(){ // update() method
    if(Target->getNewVal()==0){
        Serial.println("Recieved call to close the window");
        digitalWrite(MotorPin1,HIGH);
        digitalWrite(MotorPin2,LOW);
    }
    if(Target->getNewVal()==100){
        Serial.println("Recieved call to open the window");
        digitalWrite(MotorPin1,LOW);
        digitalWrite(MotorPin2,HIGH);
    }
    
    if(Hold->getNewVal()==1){
        Serial.println("Recieved call to hold the current position of the window");
        State->setVal(2);
        digitalWrite(MotorPin1,LOW);
        digitalWrite(MotorPin2,LOW);
    }
    
    return(true);                       // return true
  
  } // update

  void loop(){  // loop() method
    if(Current->getVal()==Target->getVal()){ // if current-state matches target-state there is nothing do -- exit loop()
      return;
    }
    
    if(Target->timeVal()>3000){
      if(Target->getVal()==0){
        Serial.println("Window closed");
      }
      if(Target->getVal()==100){
        Serial.println("Window opened");
      }
      digitalWrite(MotorPin1,LOW);
      digitalWrite(MotorPin2,LOW);
      Current->setVal(Target->getVal());      // set the current-state to the target-state
    }
  } // loop

};

////////////////////////////

void setup() {
  Serial.begin(115200);
  homeSpan.setStatusPin(2);
  //homeSpan.setPairingCode("46637726");
  homeSpan.setSketchVersion("1.0.0");
  homeSpan.setQRID("WNDO");
  homeSpan.enableOTA(); // password is "homespan-ota"
  homeSpan.begin(Category::Bridges,"Windows Bridge");
  new SpanAccessory();  
    new Service::AccessoryInformation();
      new Characteristic::Identify();

  new SpanAccessory();
    new Service::AccessoryInformation();
      new Characteristic::Identify();
      new Characteristic::Name("Window 1");
    new WindowOpener(16,17); // Create a Window Opener. Numers represent the tho GPIO for the driving relays

  new SpanAccessory();
    new Service::AccessoryInformation();
      new Characteristic::Identify();
      new Characteristic::Name("Window 2");
    new WindowOpener(18,19); // Create a Window Opener. Numers represent the tho GPIO for the driving relays

  new SpanAccessory();
    new Service::AccessoryInformation();
      new Characteristic::Identify();
      new Characteristic::Name("Window 3");
    new WindowOpener(21,22); // Create a Window Opener. Numers represent the tho GPIO for the driving relays
  /*
  new SpanAccessory();
    new Service::AccessoryInformation();
      new Characteristic::Identify();
      new Characteristic::Name("Window 4");
    new WindowOpener(23,11); // Create a Window Opener. Numers represent the tho GPIO for the driving relays
  */
} // end of setup()

void loop(){
  homeSpan.poll();
} // end of loop()
```

And this is where the esp32 keeps on resetting over and over:


```
************************************************************
Welcome to HomeSpan!
Apple HomeKit for the Espressif ESP-32 WROOM and Arduino IDE
************************************************************

** Please ensure serial monitor is set to transmit <newlines>

Message Logs:     Level 0
Status LED:       Pin 2
Device Control:   Pin -  *** WARNING: Device Control Pin is UNDEFINED
Sketch Version:   1.0.0
HomeSpan Version: 1.5.1
Arduino-ESP Ver.: 2.0.3
ESP-IDF Version:  4.4.1
ESP32 Chip:       ESP32-D0WDQ5 Rev 1 dual-core 4MB Flash
ESP32 Board:      esp32
PWM Resources:    16 channels, 8 timers, max 20-bit duty resolution
Sodium Version:   1.0.12-idf  Lib 9.4
MbedTLS Version:  mbed TLS 2.28.0
Sketch Compiled:  Jul  2 2022 07:22:09
Partition:        app0

Device Name:      Windows Bridge

Configuring Window Opener
Configuring Window Opener
Configuring Window Opener
Configuring Window Opener
ets Jun  8 2016 00:22:57

rst:0x8 (TG1WDT_SYS_RESET),boot:0x13 (SPI_FAST_FLASH_BOOT)
configsip: 0, SPIWP:0xee
clk_drv:0x00,q_drv:0x00,d_drv:0x00,cs0_drv:0x00,hd_drv:0x00,wp_drv:0x00
mode:DIO, clock div:2
load:0x3fff0030,len:1184
load:0x40078000,len:12784
load:0x40080400,len:3032
entry 0x400805e4

************************************************************
Welcome to HomeSpan!
Apple HomeKit for the Espressif ESP-32 WROOM and Arduino IDE
************************************************************
```

**Reply by HomeSpan** [reply]
I don't think you can use pin 11 on that specific board - it is connected to flash SPI.  

Is there actually a pin 11 wired on your board?

**Reply by blaisebarrette** [reply]
Well spotted! Thank you so much for taking the time looking through my code! All four windows are working rock steady!
Sorry if I wasted your time with this stupid mistake!

**Comment by HomeSpan** [comment]
No worries - really glad you're finding the library useful.

**Comment by blaisebarrette** [comment]
Quick question...

By adding the 'HoldPosition' characteristic to my Window service, shouldn't there be a 'hold' button added to the interface in the Home app?

Thanks!

**Reply by HomeSpan** [reply]
I've never gotten this to work either.  I think the hold position has been deprecated by Apple in favor of the current/target position model.  

To get around this in my own implementation of a patio screen I have a physical button on the device that, when pressed, updates the target position to match the current position, which in turn has the affect of stopping the screen at its current place. 

But unfortunately you can't do this from within the Home App or via Siri (at least not that I have been able to figure out).

**Comment by blaisebarrette** [comment]
Ok, thanks! So I won’t waist any more time on this. Opened or closed will be fine! 😉
The added control would have been nice but we probably wouldn’t have used it much anyway. 

I said it already but let me say it again, I’m having a great time coding with your library!!!

you did an amazing job!!!

**Comment by Le-Duc0069** [comment]
> #include "HomeSpan.h" 
> 
> struct WindowOpener : Service::Window { // Window Opener
>     int MotorPin1, MotorPin2;
>     SpanCharacteristic *Current;        // reference to the Current window State Characteristic
>     SpanCharacteristic *Target;         // reference to the Target window State Characteristic 
>     SpanCharacteristic *State;          // reference to the position state Characteristic
>     SpanCharacteristic *Hold;           // reference to the hold position Characteristic
>     SpanCharacteristic *Obstruction;    // reference to the Obstruction Detected Characteristic
> 
>     WindowOpener(int MotorPin1, int MotorPin2) : Service::Window(){ // constructor() method
>         
>     Current=new Characteristic::CurrentPosition(0);          // initial value of 0 means closed
>     Target=new Characteristic::TargetPosition(0);            // initial value of 0 means closed
>     State=new Characteristic::PositionState(2);              // initial value of 1 means closed
>     Hold=new Characteristic::HoldPosition(false);                // initial value of 1 means closed
>     Obstruction=new Characteristic::ObstructionDetected(false);  // initial value of 1 means closed
>         
>     Serial.print("Configuring Window Opener");   // initialization message
>     Serial.print("\n");
> 
>     this->MotorPin1=MotorPin1;
>     this->MotorPin2=MotorPin2;
> 
>     pinMode(MotorPin1,OUTPUT);
>     pinMode(MotorPin2,OUTPUT);
> 
>     digitalWrite(MotorPin1,HIGH);
>     digitalWrite(MotorPin2,HIGH);
> 
> } // end constructor
> 
>   boolean update(){ // update() method
> 
>     // see HAP Documentation for details on what each value represents
>     if(Target->getNewVal()==0){         // if the target-state value is set to 0, HomeKit is requesting the door to be in open position
>       LOG1("Opening Window\n");
>       Current->setVal(2);               // set the current-state value to 2, which means "opening"
>       digitalWrite(MotorPin1,HIGH);
>       digitalWrite(MotorPin2,LOW);
>       Obstruction->setVal(false);       // clear any prior obstruction detection
>     } else {
>       LOG1("Closing Window\n");         // else the target-state value is set to 1, and HomeKit is requesting the door to be in the closed position
>       Current->setVal(3);               // set the current-state value to 3, which means "closing"  
>       digitalWrite(MotorPin1,LOW);
>       digitalWrite(MotorPin2,HIGH);       
>       Obstruction->setVal(false);       // clear any prior obstruction detection
>     }
> 
>     if(State->getNewVal()==0){
> 
>     }
>     
>     return(true);                       // return true
>   
>   } // update
> 
>   void loop(){  // loop() method
> 
>     if(Current->getVal()==Target->getVal()) // if current-state matches target-state there is nothing do -- exit loop()
>     digitalWrite(MotorPin1,LOW);
>     digitalWrite(MotorPin2,LOW);
>       return;
> 
>     if(Current->getVal()==4)                // if the current-state is stopped, there is nothing more to do - exit loop()     
>     digitalWrite(MotorPin1,LOW);
>     digitalWrite(MotorPin2,LOW);
>       return;
> 
>     Current->setVal(Target->getVal());      // set the current-state to the target-state
>        
>   } // loop
>   
> };
> 
> 
> void setup() {
>   Serial.begin(115200);
>   homeSpan.setStatusPin(2);
>   homeSpan.setPairingCode("46637727");
>   homeSpan.setSketchVersion("1.0.0");
>   homeSpan.setQRID("WNDO");
>   homeSpan.enableOTA(); // password is "homespan-ota"
>   homeSpan.begin(Category::Windows,"Windows control");
>   new SpanAccessory();
>     new Service::AccessoryInformation();
>       new Characteristic::Identify(); 
>     new WindowOpener(16,17); // Create a Window Opener (see DEV_DoorsWindows.h for definition)
> } // end of setup()
> 
> void loop(){
>   homeSpan.poll();
> } // end of loop()

can you give me the installation diagram. THANK YOU
